### DISCUSSION TITLE
Joint 3 of both arms in the iCub_SIM will not move if 15.385 degrees is commanded

### DISCUSSION THREAD

**Discussion by gccunha1** [robotology/community#442] [Q&A] [answered]
## Brief description
Hello!
I noticed that joint number 3 in the iCub_SIM right arm will not move if the reference is the minimum allowed value, which is 15.385 degrees. This happens both using code, as well as using yarpmotorgui.

## Detailed context
I previously posted issue #406 here about adding fiducial marker textures to the iCub_SIM and place them on the robot's hand.
I noticed that sometimes the hand and the markers would be in completely different positions, but thought I was the one miscalculating the world coordinates, but then realized this happened when a value close to 15.285 degrees was commanded to joint 3 of the right arm, and it ended up not moving. This happens with joint 3 of the left arm aswell.

I attached a .zip file containing a video of what happens in yarpmotorgui, which is identical when commands are sent using code.

## Accompanying material
[Qt Robot Motor GUI V2.0 2020-11-05 16-06-23.zip](https://github.com/robotology/QA/files/5495556/Qt.Robot.Motor.GUI.V2.0.2020-11-05.16-06-23.zip)

**Comment by pattacini** [comment, marked as answer]
Hi @gccunha1 

I've made a test with a command line swiss knife called `simpleClient`.

```sh
>> simpleClient --robot icubSim --part right_arm
-> get llim 3
llim: limits: (15.50 106.00)
-> set pos 3 15.50
pos: moving 3 to 15.50 
-> get encs
Bottle: get encs
encs: (-25.00 20.00 0.00 15.50 0.00 -0.00 0.00 59.00 20.00 20.00 20.00 10.00 10.00 10.00 10.00 10.00 )
-> set pos 3 106.0
Bottle: set pos 3 106.0
pos: moving 3 to 106.00
-> get encs
Bottle: get encs
encs: (-25.00 20.00 0.00 106.00 0.00 0.00 0.00 59.00 20.00 20.00 20.00 10.00 10.00 10.00 10.00 10.00 )
```

As you can see from the log above, the actual joint limits are `(15.5 106)` degrees. That can explain why you've been experiencing problems with the `yarpmotorgui`, which instead is wrongly allowing you to play with different bounds, apparently.

When it comes down to the code, you must have asked to move to `15.385` deg without checking whether this target was feasible, i.e. within the allowed range. To this end, rely on [`IControlLimits::getLimits()`][1].

Regarding the weird beahavior of `yarpmotorgui`, my suggestion is to open up an issue in YARP at https://github.com/robotology/yarp/issues. You may want to link to this issue as a reference.

[1]: http://yarp.it/git-master/classyarp_1_1dev_1_1IControlLimits.html#a67727bdb91bddeb1c070d44904d501cc

**Reply by jimingda** [reply]
Hi Ugo, where can i get info or guide on this simpleClient ?

**Reply by jimingda** [reply]
and when i followed your procedure above, i didn't get degree info like "encs: (-25.00 20.00 0.00 ... "
after command  "gen encs  ",   as followed

```console
root@db61dac93e4e:~# simpleClient --robot icubSim --part right_arm
(device remote_controlboard) (local "/icubSim/right_arm/simpleclient") (part right_arm) (remote "/icubSim/right_arm") (robot icubSim)[DEBUG] |yarp.dev.PolyDriver|remote_controlboard| Parameters are (device remote_controlboard) (local "/icubSim/right_arm/simpleclient") (part right_arm) (remote "/icubSim/right_arm") (robot icubSim)
[INFO] |yarp.os.Port|/icubSim/right_arm/simpleclient/rpc:o| Port /icubSim/right_arm/simpleclient/rpc:o active at tcp://172.17.0.2:10081/
[INFO] |yarp.os.Port|/icubSim/right_arm/simpleclient/command:o| Port /icubSim/right_arm/simpleclient/command:o active at tcp://172.17.0.2:10082/
[INFO] |yarp.os.Port|/icubSim/right_arm/simpleclient/stateExt:i| Port /icubSim/right_arm/simpleclient/stateExt:i active at tcp://172.17.0.2:10083/
[INFO] |yarp.os.impl.PortCoreOutputUnit|/icubSim/right_arm/simpleclient/rpc:o| Sending output from /icubSim/right_arm/simpleclient/rpc:o to /icubSim/right_arm/rpc:i using tcp
[INFO] |yarp.os.impl.PortCoreOutputUnit|/icubSim/right_arm/simpleclient/command:o| Sending output from /icubSim/right_arm/simpleclient/command:o to /icubSim/right_arm/command:i using udp
[INFO] |yarp.os.impl.PortCoreInputUnit|/icubSim/right_arm/simpleclient/stateExt:i| Receiving input from /icubSim/right_arm/stateExt:o to /icubSim/right_arm/simpleclient/stateExt:i using udp
[INFO] |yarp.dev.PolyDriver|remote_controlboard| Created device <remote_controlboard>. See C++ class RemoteControlBoard for documentation.
Working with 16 axes
Device active...
-> get llime 3
Bottle: get llime 3
llim: limits: (15.00 106.00)
-> set pos 3 15.50
Bottle: set pos 3 15.5
pos: moving 3 to 15.50
-> gen encs     
Bottle: gen encs
-> set pos 3 106
Bottle: set pos 3 106
pos: moving 3 to 106.00
-> gen encs
Bottle: gen encs
->
```

**Reply by pattacini** [reply]
Hi @jimingda

>where can i get info or guide on this simpleClient ?

The `simpleClient` module is a tool we do use only for debugging and as such is not really documented. You may try to issue `help` from its terminal interface to get a list of the available commands, although it won't help you that much, I suppose. In essence, you'll find out those low-level commands that are sent out through YARP ports when from the C++ code the user calls for the standard motor API services.

>and when i followed your procedure above, i didn't get degree info like "encs: (-25.00 20.00 0.00 ... "
after command "gen encs "

This is because there's most likely a typo in the command you typed in: it's `get encs` and not `gen encs`.

**Comment by gccunha1** [comment]
> When it comes down to the code, you must have asked to move to 15.385 deg without checking whether this target was feasible, i.e. within the allowed range. To this end, rely on IControlLimits::getLimits()

Yes, I was not doing that and I actually didn't know that function, I was relying on the values from yarpmotorgui and I hard-coded the bounds. I will open the issue at  https://github.com/robotology/yarp/issues.

Thanks for the reply!
